modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/dao/es/ElasticSearchDetectorMappingService.java

```diff
@@ -0,0 +1,377 @@
+        } catch (IOException e) {
+            this.exceptionCount.inc();
+            log.error("Error ES lookup", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public String createDetectorMapping(CreateDetectorMappingRequest createRequest) {
+        Set<String> fields = getFields(createRequest);
+        Set<String> newFieldMappings = getFieldsWithoutExistingMapping(fields);
+        updateIndexMappings(newFieldMappings);
+        final DetectorMappingEntity detectorMappingEntity = new DetectorMappingEntity(createRequest.getUser(),
+            createRequest.getDetector(), QueryUtil.buildQuery(createRequest.getExpression()), true,
+            System.currentTimeMillis(), System.currentTimeMillis());
+        final IndexRequest indexRequest = new IndexRequest(elasticSearchConfig.getIndexName(),
+            elasticSearchConfig.getDocType());
+        indexRequest.source(DetectorMappingEntityToJson(detectorMappingEntity), XContentType.JSON);
+        try {
+            IndexResponse indexResponse = elasticSearchClient.index(indexRequest, RequestOptions.DEFAULT);
+            return indexResponse.getId();
+        } catch (IOException e) {
+            log.error(String.format("Indexing mapping %s failed", DetectorMappingEntityToJson(detectorMappingEntity), e));
+            throw new RuntimeException(e);
+        }
+    }
+    
+    private void updateIndexMappings(Set<String> newFieldMappings) {
+        PutMappingRequest request = new PutMappingRequest(elasticSearchConfig.getIndexName());
+        Map<String, Object> jsonMap = new HashMap<>();
+        Map<String, Object> type = new HashMap<>();
+        type.put("type", "keyword");
+        Map<String, Object> properties = new HashMap<>();
+        newFieldMappings.forEach(field -> {
+            properties.put(field, type);
+        });
+        jsonMap.put("properties", properties);
+        request.source(jsonMap);
+        request.type(elasticSearchConfig.getDocType());
+        try {
+            elasticSearchClient.indices().putMapping(request, RequestOptions.DEFAULT);
+        } catch (IOException e) {
+            log.error("Error updating mappings", e);
+            throw new RuntimeException(e);
+        }
+    }
+    
+    private Set<String> getFieldsWithoutExistingMapping(Set<String> fields) {
+        GetMappingsRequest request = new GetMappingsRequest();
+        request.indices(elasticSearchConfig.getIndexName());
+        
+        try {
+            GetMappingsResponse mappingsResponse = elasticSearchClient.indices().getMapping(request, RequestOptions.DEFAULT);
+            Set<String> mappeddFields = ((Map<String, String>)mappingsResponse.getMappings()
+                    .get(elasticSearchConfig.getIndexName())
+                    .get(elasticSearchConfig.getDocType()).sourceAsMap().get("properties")).entrySet().stream()
+                    .map(en -> en.getKey()).collect(Collectors.toSet());
```

Comment:
Use `mappedFields` instead of `mappeddFields`.

Also there is a lot going on in this single line of code. Consider breaking it up a bit and using formatting to clarify what's going on.

Finally, the method is called `getFieldsWithoutExistingMapping()`, which makes it sound like this method has no side effects on the passed `fields` set since `getXXX()` methods are usually understood not to modify anything. In this case you are removing fields from that set, which is a side effect. Consider renaming this to something like `removeFieldsHavingExistingMapping()`.

Alternatively, you can leave the original set intact, and just create a new set. That would remove the side effect from `getFieldsWithoutExistingMapping()` and allow you to keep the same name.

Reply:
Updated.